streams/src/main/java/org/apache/kafka/streams/kstream/internals/KeyValueWithTimestampStoreMaterializer.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.DefaultRecordConverter;
+import org.apache.kafka.streams.state.KeyValueBytesStoreSupplier;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.RecordConverter;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.Stores;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.internals.KeyValueIteratorFacade;
+import org.apache.kafka.streams.state.internals.KeyValueStoreBuilder;
+import org.apache.kafka.streams.state.internals.WrappedStateStore;
+
+import java.util.List;
+import java.util.Map;
+
+public class KeyValueWithTimestampStoreMaterializer<K, V> {
+    private final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materialized;
+
+    public KeyValueWithTimestampStoreMaterializer(final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materialized) {
+        this.materialized = materialized;
+    }
+
+    /**
+     * @return  StoreBuilder
+     */
+    public StoreBuilder<KeyValueStore<K, V>> materialize() {
+        KeyValueBytesStoreSupplier supplier = (KeyValueBytesStoreSupplier) materialized.storeSupplier();
```

Comment:
This looks a bit suspicious to me: users could also call `Stores.persistentKeyValueWithTimestampStore` and passed in the returned object o `Materialized`, in which case `V` is already a `ValueAndTimestamp<V1>` right?

Reply:
Yes. In fact, we expect the supplier to return a store that can handle `ValueAndTimestamp`.

However, we need the return type of the supplier to be a byte store -- and thus, we cannot distinguish between non-timestamp-byte-stores and timestamped-byte-store easily. However, we cover this case at an outer level and wrap a plain-key-value store with a proxy if a RecordConverter is not implemented (was is only an indicator)

In the end, because we want all most-inner stores to be plain-byte stores, we can't distinguish them -- however, if we exclude the upgrade of an existing store into a new store, both work work for both cases because both only take bytes, and only use the key for lookups -- basically both stores are agnostic is the value bytes are the one or the other.